airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/S3ConsumerFactory.kt

```diff
@@ -208,56 +251,78 @@ class S3ConsumerFactory {
+            )
+        }
+        if (writeConfig.minimumGenerationId != writeConfig.generationId) {
+            throw IllegalArgumentException("Hybrid refreshes are not yet supported.")
+        }
+
+        // This is truncate sync and try to determine if the current generation
+        // data is already present
+        val namespace = writeConfig.namespace
+        val stream = writeConfig.streamName
+        val outputBucketPath = writeConfig.outputBucketPath
+        val pathFormat = writeConfig.pathFormat
+        // generationId is missing, assume the last sync was ran in non-resumeable refresh
+        // mode,
+        // cleanup files
+        val currentGenerationId =
+            storageOperations.getStageGeneration(namespace, stream, outputBucketPath, pathFormat)
+        var filterByCurrentGen = false
+        if (currentGenerationId != null) {
+            // if minGen = gen = retrievedGen and skip clean up
+            val hasDataFromCurrentGeneration = currentGenerationId == writeConfig.generationId
+            if (hasDataFromCurrentGeneration) {
+                LOGGER.info {
+                    "Preserving data from previous sync for stream ${writeConfig.streamName} since it matches the current generation ${writeConfig.generationId}"
                 }
-                // if minGen = gen = retrievedGen and skip clean up
-                val hasDataFromCurrentGeneration = currentGenerationId == writeConfig.generationId
-                if (hasDataFromCurrentGeneration) {
-                    LOGGER.info {
-                        "Preserving data from previous sync for stream ${writeConfig.streamName} since it matches the current generation ${writeConfig.generationId}"
-                    }
-                } else {
-                    LOGGER.info {
-                        "No data exists from previous sync for stream ${writeConfig.streamName} from current generation ${writeConfig.generationId}, " +
-                            "proceeding to clean up existing data"
-                    }
+                // There could be data dangling from T-2 sync if current generation failed in T-1
+                // sync.
+                filterByCurrentGen = true
+            } else {
+                LOGGER.info {
+                    "No data exists from previous sync for stream ${writeConfig.streamName} from current generation ${writeConfig.generationId}, " +
+                        "proceeding to clean up existing data"
                 }
-                return !hasDataFromCurrentGeneration
             }
-            else -> {
-                throw IllegalArgumentException("Hybrid refreshes are not yet supported.")
+        } else {
+            LOGGER.info {
+                "Missing generationId from the lastModified object, proceeding with cleanup for stream ${writeConfig.streamName}"
             }
         }
+
+        if (filterByCurrentGen) {
```

Comment:
why not use the sentinel value?
```kotlin
return storageOps.list(
  ...,
  generationId = if (filterByCurrentGen) { writeConfig.generationId } else { null }
)
```

Reply:
No reason, changes were incrementally done so at some point sentinel value wasn't there